File: 238-product-of-array-except-self/238. Product of Array Except Self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        left_multiply = []
        right_multiply = []
        ans = []
        total = 1
        #left
        for num in nums:
            total = num*total
            left_multiply.append(total)
        #right
        total = 1
        for num in reversed(nums):
            total = num*total
            right_multiply.append(total)
        right_multiply.reverse()

        ans = [right_multiply[1]]
        for i in range(1,len(nums)-1):
            ans.append(left_multiply[i-1]*right_multiply[i+1])
        ans += [left_multiply[-2]]
        
        print(left_multiply)
        print(right_multiply)
        return ans

File: 238-product-of-array-except-self/238. Product of Array Except Self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> lprod;
        vector<int> rprod;
        vector<int> ans;
        lprod.push_back(1);
        lprod.push_back(nums[0]);
        rprod.push_back(nums.back());

        for(int i=1;i<nums.size();i++){
            lprod.push_back(nums[i]*lprod.back());
        }

        for(int i=nums.size()-2;i>=0;i--){
            rprod.push_back(nums[i]*rprod.back());
        }
        std::reverse(rprod.begin(),rprod.end());
        rprod.push_back(1);

        // for(int num: lprod){
        //     cout << num << " ";
        // }
        // cout << endl;
        // for(int num: rprod){
        //     cout << num << " ";
        // }
        // cout << endl;

        for(int i=0;i<nums.size();i++){
            ans.push_back(lprod[i]*rprod[i+1]);
        }
        return ans;
    }
};